cpp/src/barretenberg/stdlib/primitives/field/field.cpp

```diff
@@ -1159,27 +1189,37 @@ std::vector<bool_t<ComposerContext>> field_t<ComposerContext>::decompose_into_bi
         y_lo += shifted_high_limb;
         y_lo.normalize();
 
-        // A carry was necessary if and only if the 128th bit y_lo_hi of y_lo is 0.
-        auto [y_lo_lo, y_lo_hi, zeros] = y_lo.slice(128, 128);
-        // This copy constraint, along with the constraints of field_t::slice, imposes that y_lo has bit length 129.
-        // Since the integer y_lo is at least -2**128+1, which has more than 129 bits in `Fr`, the implicit range
-        // constraint shows that y_lo is non-negative.
-        context->assert_equal(
-            zeros.witness_index, context->zero_idx, "field_t: bit decomposition_fails: high limb non-zero");
-        // y_borrow is the boolean "a carry was necessary"
-        field_t<ComposerContext> y_borrow = -(y_lo_hi - 1);
-        // If a carry was necessary, subtract that carry from p_hi
-        // y_hi = (p_hi - y_borrow) - sum_hi
-        field_t<ComposerContext> y_hi = -(shifted_high_limb / shift) + (p_hi);
-        y_hi -= y_borrow;
-        // As before, except that now the range constraint is explicit, this shows that y_hi is non-negative.
-        y_hi.create_range_constraint(128, "field_t: bit decomposition fails: y_hi is too large.");
+        if constexpr (IsSimulator<ComposerContext>) {
```

Comment:
This is perhaps the most complicated new "circuit" code. We have to check that the purported bit decomposition gives a reduced (i.e. less-than-modulus) uint. IMO it adds clarity here in the broader context of the function--the fact that the simulator test passes illustrates that the more complicated circuit logic does indeed reflect what is done in the comments of the function (the simulator path uses the same notation as the comments).